homeassistant/components/asuswrt/device_tracker.py

```diff
@@ -1,70 +1,144 @@
-
-    async def async_connect(self):
-        """Initialize connection to the router."""
-        # Test the router is accessible.
-        data = await self.connection.async_get_connected_devices()
-        self.success_init = data is not None
-
-    async def async_scan_devices(self):
-        """Scan for new devices and return a list with found device IDs."""
-        await self.async_update_info()
-        return list(self.last_results)
-
-    async def async_get_device_name(self, device):
-        """Return the name of the given device or None if we don't know."""
-        if device not in self.last_results:
-            return None
-        return self.last_results[device].name
-
-    async def async_update_info(self):
-        """Ensure the information from the ASUSWRT router is up to date.
-
-        Return boolean if scanning successful.
-        """
-        _LOGGER.debug("Checking Devices")
-
-        try:
-            self.last_results = await self.connection.async_get_connected_devices()
-            if self._connect_error:
-                self._connect_error = False
-                _LOGGER.info("Reconnected to ASUS router for device update")
-
-        except OSError as err:
-            if not self._connect_error:
-                self._connect_error = True
-                _LOGGER.error(
-                    "Error connecting to ASUS router for device update: %s", err
-                )
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up device tracker for AsusWrt component."""
+    router = hass.data[DOMAIN][entry.entry_id][DATA_ASUSWRT]
+    tracked = set()
+
+    @callback
+    def update_router():
+        """Update the values of the router."""
+        add_entities(router, async_add_entities, tracked)
+
+    router.async_on_close(
+        async_dispatcher_connect(hass, router.signal_device_new, update_router)
+    )
+
+    update_router()
+
+
+@callback
+def add_entities(router, async_add_entities, tracked):
+    """Add new tracker entities from the router."""
+    new_tracked = []
+
+    for mac, device in router.devices.items():
+        if mac in tracked:
+            continue
+
+        new_tracked.append(AsusWrtDevice(router, device))
+        tracked.add(mac)
+
+    if new_tracked:
+        async_add_entities(new_tracked, True)
+
+
+class AsusWrtDevice(ScannerEntity):
+    """Representation of a AsusWrt device."""
+
+    def __init__(self, router: AsusWrtRouter, device) -> None:
+        """Initialize a AsusWrt device."""
+        self._router = router
+        self._mac = device.mac
+        self._name = device.name or DEFAULT_DEVICE_NAME
+        self._active = False
+        self._icon = None
+        self._attrs = {}
+
+    @callback
+    def async_update_state(self) -> None:
+        """Update the AsusWrt device."""
+        device = self._router.devices[self._mac]
+        self._name = device.name or DEFAULT_DEVICE_NAME
```

Comment:
Updating the name won't have any dynamic effect since we set a unique id. The entity is registered in the entity registry before `async_added_to_hass` is called. The updated name will be used though if the user updates the entity in the entity registry. Then the entity will be removed and readded with the updated name as default name.

Just an FYI.

Reply:
Removed, also because device.name is updated only once when detected